### PR TITLE
[IO] Use a unique, temporary name while writing a file to cache

### PR DESCRIPTION
**Pull request by eguiraud**
This way, if another process requests the file from the cache while
the file is still being written/downloaded, it does not "find" it
and does not read an incomplete file.

This should fix failures such as [these](https://github.com/root-project/root/pull/9495#issuecomment-1005637703) where multiple process (in this case multiple ROOT tests) require to cache the same file concurrently.

### CODE REVIEW

File: io/io/src/TFile.cxx [hunk 1]
```diff
@@ -3939,13 +3938,19 @@ TFile *TFile::OpenFromCache(const char *name, Option_t *, const char *ftitle,
...
             fgCacheFileForce = kFALSE;
-            if (need2copy && !TFile::Cp(name, cachefilepath)) {
-               ::Warning("TFile::OpenFromCache", "you want to read through a cache, but I "
-                         "cannot make a cache copy of %s - CACHEREAD disabled",
-                         cachefilepathbasedir.Data());
-               fgCacheFileForce = forcedcache;
-               if (fgOpenTimeout != 0)
-                  return nullptr;
+            if (need2copy) {
+               TUUID uuid;
```
Comment (pcanal) [inline io/io/src/TFile.cxx:3942]: Since this is a transient name.  `gSystem->GetPID()` will be sufficient and **might** be faster than TUUID.

Comment (eguiraud) [reply io/io/src/TFile.cxx]: done

**Comment by eguiraud** [comment]
@phsft-bot build

**Review by Axel-Naumann — APPROVED** [review]
LGTM thanks for doing this!

**Comment by eguiraud** [comment]
Mmmh the outer if/else logic is wrong now (we need to take the `else` branch if `TFile::Cp` succeeds). I'm on it.

**Comment by eguiraud** [comment]
I tested it locally and it seems to do the job now. If two processes get to the creation of the cache directory hierarchy precisely at the same time, it seems that there can still be a race there where one process fails to create the cache directory but it also doesn't see the full hierarchy already there. In that case there is no hard failure though: the process that fails to create the cache directory simply reads the file remotely rather than caching it.

**Review by pcanal — APPROVED** [review]

**Pull request merged**